docs/src/service-workers.md

```diff
@@ -0,0 +1,349 @@
+
+**If your app uses Service Workers**, here's the scenarios that Playwright supports:
+
+1. Testing the page exactly like a user would experience it. This works out of the box in all supported browsers.
+1. Test your page without a Service Worker. Set [`option: Browser.newContext.serviceWorkers`] to `'block'`. You can test your page as if no Service Worker was registered.
+1. Listen for and route network traffic via Playwright, whether it comes from a Service Worker or not. In Firefox and WebKit, set [`option: Browser.newContext.serviceWorkers`] to `'block'` to avoid Service Worker network traffic entirely. In Chromium, either block Service Workers or use [`method: BrowserContext.route`].
+1. (Chromium-only) Test your Service Worker implementation itself. Use [`method: BrowserContext.serviceWorkers`] to get access to the Service Worker and evaluate there.
+
+## Disabling Service Workers
+
+Set [`option: Browser.newContext.serviceWorkers`] to `'block'`.
+
+
+## Service Worker Fetch
+
+:::note
+The next sections are only currently supported when using Playwright with Chrome/Chromium. In Firefox and WebKit, if a Service Worker has a FetchEvent handler, Playwright will **not** emit Network events for all network traffic.
+:::
+
+### Accessing Service Workers and Waiting for Activation
+
+You can use [`method: BrowserContext.serviceWorkers`] to list the Service [Worker]s, or specifically watch for the Service [Worker] if you anticipate a page will trigger its [registration](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerContainer/register):
+
+```js tab=js-ts
+const [ serviceworker ] = await Promise.all([
+  context.waitForEvent('serviceworker'),
+  page.goto('/example-with-a-service-worker.html'),
+]);
+```
+
+```js tab=js-js
+const [ serviceworker ] = await Promise.all([
+  context.waitForEvent('serviceworker'),
+  page.goto('/example-with-a-service-worker.html'),
+]);
+```
+
+```python async
+async with context.expect_event("serviceworker") as event_info:
+    await page.goto('/example-with-a-service-worker.html')
+serviceworker = await event_info.value
+```
+
+```python sync
+with context.expect_event("serviceworker") as event_info:
+    page.goto('/example-with-a-service-worker.html')
+serviceworker = event_info.value
+```
+
+```csharp
+var waitForServiceWorkerTask = page.WaitForServiceWorkerAsync();
+await page.GotoAsync('/example-with-a-service-worker.html');
+var serviceworker = await waitForServiceWorkerTask;
+```
+
+[`event: BrowserContext.serviceWorker`] is fired ***before*** the Service Worker's main script has been evaluated, so ***before*** calling service[`method: Worker.evaluate`] you should wait on its activation:
+
+```js tab=js-ts
+await expect.poll(() => serviceworker.evaluate(() => (self as any).registration.active?.state)).toBe('activated');
+```
+
+```js tab=js-js
+await expect.poll(() => serviceworker.evaluate(() => self.registration.active?.state)).toBe('activated');
+```
```

Comment:
Post-landing the doc change, let's sort out an API to add to do this idiomatically. I don't like poll to begin with, and Python/NET/Java don't seem to have the equivalent.

If we want to hold of on introducing an API, but still want to provide a workaround, we can share the activationPromise strategy on Page instead of Service Worker.